### PR TITLE
WIP / Remove old mongo

### LINKED ISSUE #1940: Deprecate old-Mongo-connector-implementation and switch to new-Mongo-connector
Based on QScript etc.

### PR DESCRIPTION
**Pull request by sellout**
Fixes #1940 (and others to be enumerated).

### CODE REVIEW

**Comment by jdegoes** [comment]
🎉

**Comment by sellout** [comment]
Here are integration tests that have regressed from old Mongo, but are _likely_ caused by problems in QScript (or earlier). I say “likely” because they are the tests that don’t work for _any_ QScript connectors. Fixing these should probably be a priority over any other QScript issues, since getting new Mongo up to par with old Mongo is probably more important than adding new features to other connectors.

- [ ] caseVariations
- [ ] caseWithGroup
- [ ] citiesByLargestZip
- [ ] different-flattens
- [ ] flattenGroupedObject
- [ ] groupByFlatten
- [ ] joins/groupedJoin
- [ ] joins/nonJsJoinCondition
- [ ] joins/symmetricNonJsJoinCondition
- [ ] locationCount
- [ ] temporal/filterOnDate
- [ ] multistageFlatten
- [ ] multitypeFlatten
- [ ] shortCities
- [ ] skipCount
- [ ] unifyFlattenedFieldsWithMultipleUses
- [ ] unifyFlattenedFieldsWithUnflattened

**Comment by alissapajer** [comment]
@sellout a handful of those integration tests are in theory skipped due to bug #1951 / #1929 (qscript provenance bug). Is that bug fixed in this PR? In other words, is that still the reason those tests aren't passing or is there another problem?

**Comment by alissapajer** [comment]
@sellout could you add "closes" notes to the PR description for all the issues that this PR should close?

**Comment by alissapajer** [comment]
@sellout did any integration tests come to life when you fixed #1755? (If that fix is contained in this branch?)

**Comment by sellout** [comment]
@alissapajer I have the Provenance work on a separate branch, so the ones blocked by #1929 aren’t yet fixed on this branch. And yeah, there are tests that were fixed by #1755. I’ll go through all my “on hold” tickets and update the description when I have a few more things ironed out.

**Comment by alissapajer** [comment]
@sellout I updated your list of integration tests with all the issues that already exist. It's possible that some of those tests a second underlying issue, so we can continue to update those ticket numbers as necessary.

**Comment by sellout** [comment]
@alissapajer Thanks!

**Comment by alissapajer** [comment]
@sellout in the above comment that lists the integration test that are likely qscript problems, when I think I've addressed a test so that it's no longer a qscript problem (e.g. multiple connectors have the test passing) I will check off the box.

File: mongodb/src/main/scala/quasar/physical/mongodb/workflowbuilder.scala [hunk 1]
```diff
@@ -340,465 +320,244 @@ object WorkflowBuilder {
...
-            sb.toJs.map { splice =>
-              CollectionBuilderF(
-                chain(wf,
-                  $simpleMap[F](NonEmptyList(MapExpr(JsFn(jsBase, (base.toDocVar.toJs >>> splice)(jscore.Ident(jsBase))))), ListMap())),
-                Root(),
-                None)
-            })
-        }
-    }
+  // TODO: Determine if this is necessary. If so, flesh it out.
```
Comment (rintcius) [inline mongodb/src/main/scala/quasar/physical/mongodb/workflowbuilder.scala]: In what sense can this be unnecessary? And what needs fleshing out if it's necessary?

File: mongodb/src/test/scala/quasar/physical/mongodb/expression/spec.scala [hunk 2]
```diff
@@ -150,17 +149,6 @@ class ExpressionSpec extends quasar.Qspec {
     }
   }
 
-  "toJs" should {
-    import quasar.jscore._
-
-    "handle addition with epoch date literal" in {
```
Comment (rintcius) [inline mongodb/src/test/scala/quasar/physical/mongodb/expression/spec.scala]: what's the reason for removing this?

File: mongodb/src/main/scala/quasar/physical/mongodb/planner/FuncHandler.scala [hunk 3]
```diff
@@ -342,12 +363,12 @@ object FuncHandler {
     }
 
   def handle2_6[F[_]: FuncHandler]: F ~> OptionFree[Expr2_6, ?] =
-    implicitly[FuncHandler[F]].handle2_6
```
Comment (sellout) [inline mongodb/src/main/scala/quasar/physical/mongodb/planner/FuncHandler.scala]: Yeah – the `apply` on the companion gives us this without having to use `implicitly`. And we get the `apply` for free, too, in the cases we can use `@typeclass`.

File: it/src/main/resources/tests/mongodb/distinctStar.test [hunk 4]
```diff
@@ -1,14 +1,13 @@
 {
```
Comment (sellout) [inline it/src/main/resources/tests/mongodb/distinctStar.test]: So, when we had just Mongo, and hadn’t really finalized MRA, we had some expectations about what a connector should do for a query. With QScript and other connectors, we realized some of those were wrong.

So we implemented them correctly for QScript stuff, but moved the old ones to a Mongo-specific directory to maintain until we fixed the Mongo connector to do the right thing. So, as more of Mongo gets fixed, these tests will go away and the identical test not in the Mongo directory will be enabled for Mongo.

Also, there’s a Couchbase directory that is roughly for the same purpose, but specifically to handle Couchbase’s idiosyncratic join support – again, until it’s handled the same as other connectors.

**Comment by sellout** [comment]
@rintcius There are a few reasons tests went away – the one `planner.scala` (or similar) file went away because it was planner tests for old Mongo. The `workflowbuilder.scala` tests went away because QScript now handles all the complicated bits of WorkflowBuilder that we were testing. As I mentioned in the call last Thursday, WorkflowBuilder itself _may_ go away, or it may still be useful as an IR, I’m not quite sure.

Of course, some other tests may have gone away by accident …

File: it/src/main/resources/tests/citiesByLargestZip.test [hunk 5]
```diff
@@ -1,14 +1,18 @@
...
+        "mongodb_3_0":    "pending",
+        "mongodb_3_2":    "pending",
+        "mongodb_3_4":    "pending",
+        "mongodb_read_only": "pending",
+        "postgresql":     "pending",
+        "spark_hdfs":     "pending",
+        "spark_local":    "pending"
     },
     "NB": "Skipped for all connectors due to reduce/sort ordering bug.
            Bug is fixed on @sellout's new mongo branch, but the fix breaks old mongo.
```
Comment (sellout) [inline it/src/main/resources/tests/citiesByLargestZip.test]: Yeah, I think there were a few tests with this comment – some of them have been fixed. This one is probably now hitting a different issue.

File: it/src/main/resources/tests/mongodb/temporal/filterOnDate.test [hunk 6]
```diff
@@ -1,22 +0,0 @@
...
-    "data": "../../days.data",
-    "query": "select day from `../../days`
-              where
-                ((ts > date(\"2014-08-17\") and ts <= date(\"2014-08-20\"))
-                  and ts != date(\"2014-08-19\"))
-                or ts = date(\"2014-08-22\")",
-    "predicate": "exactly",
-    "ignoreResultOrder": true,
-    "expected": ["Monday", "Wednesday", "Friday"]
-}
```
Comment (sellout) [inline it/src/main/resources/tests/mongodb/temporal/filterOnDate.test]: This is a case where Mongo now passes the non-Mongo-specific version of this test. This one basically relied on bad behavior comparing date and timestamp values.

File: mongodb/src/main/scala/quasar/physical/mongodb/expression/package.scala [hunk 7]
```diff
@@ -46,103 +44,6 @@ package object expression {
     case DocVar.ROOT(Some(tail)) => tail
   } (DocVar.ROOT(_))
 
-  // The following few cases are places where the ExprOp created from
-  // the LogicalPlan needs special handling to behave the same when
-  // converted to JS.
-  // TODO: See SD-736 for the way forward.
```
Comment (sellout) [inline mongodb/src/main/scala/quasar/physical/mongodb/expression/package.scala]: That looks like a reference to something in our Jira. Yes, we still have a Jira, but thankfully we no longer use it: https://slamdata.atlassian.net/browse/SD-736

Comment (sellout) [inline mongodb/src/main/scala/quasar/physical/mongodb/expression/package.scala]: To decode that particular one SlamScript is now QScript, and MapFunc basically solves this problem for us.

**Comment by rintcius** [comment]
@sellout I looked through the other big chunks of tests that were removed & those removals look ok too (eg StdLib tests were testing both new and old mongo conn.). So 👍

**Comment by alissapajer** [comment]
Closing this now that we have PR https://github.com/quasar-analytics/quasar/pull/2523 open against the `next-major` branch.

**Pull request closed without merging**